File: design-pattern/strategy/ex2-text_list/go/main.go
package main

import (
	"fmt"
	"strings"
)

type OutputFormat int

const (
	Markdown OutputFormat = iota
	Html
)

type ListStrategy interface {
	Start(builder *strings.Builder)
	End(builder *strings.Builder)
	AddListItem(builder *strings.Builder, item string)
}

type MarkdownListStrategy struct{}

func (m *MarkdownListStrategy) Start(builder *strings.Builder) {}
func (m *MarkdownListStrategy) End(builder *strings.Builder)   {}
func (m *MarkdownListStrategy) AddListItem(builder *strings.Builder, item string) {
	builder.WriteString(" * " + item + "\n")
}

type HtmlListStrategy struct{}

func (h *HtmlListStrategy) Start(builder *strings.Builder) {
	builder.WriteString("<ul>\n")
}
func (h *HtmlListStrategy) End(builder *strings.Builder) {
	builder.WriteString("</ul>\n")
}
func (h *HtmlListStrategy) AddListItem(builder *strings.Builder, item string) {
	builder.WriteString(" <li>" + item + "</li>\n")
}

type TextProcessor struct {
	builder      strings.Builder
	listStrategy ListStrategy
}

func NewTextProcessor(listStrategy ListStrategy) *TextProcessor {
	return &TextProcessor{strings.Builder{}, listStrategy}
}

func (t *TextProcessor) SetOutputFormat(fmt OutputFormat) {
	switch fmt {
	case Markdown:
		t.listStrategy = &MarkdownListStrategy{}
	case Html:
		t.listStrategy = &HtmlListStrategy{}
	}
}

func (t *TextProcessor) AppendList(items []string) {
	s := t.listStrategy
	s.Start(&t.builder)
	for _, item := range items {
		s.AddListItem(&t.builder, item)
	}
	s.End(&t.builder)
}

func (t *TextProcessor) Reset() {
	t.builder.Reset()
}

func (t *TextProcessor) String() string {
	return t.builder.String()
}

func main() {
	tp := NewTextProcessor(&MarkdownListStrategy{})
	tp.AppendList([]string{"foo", "Yao", "Ling"})
	fmt.Println(tp)

	tp.Reset()
	tp.SetOutputFormat(Html)
	tp.AppendList([]string{"foo", "Yao", "Ling"})
	fmt.Println(tp)
}

File: design-pattern/strategy/ex2-text_list/python/main.py
from abc import ABC, abstractmethod
from io import StringIO

list_content = StringIO()

class TextProcessor():
    def __init__(self, list_strategy):
        self._list_strategy = list_strategy

    def set_list_strategy(self, list_strategy):
        self._list_strategy = list_strategy

    def append_list(self, items=[]):
        self._list_strategy.start(list_content)
        for item in items:
            self._list_strategy.addListItem(list_content, item)
        self._list_strategy.end(list_content)

    def reset(self):
        list_content.truncate(0)
        list_content.seek(0)

    def output(self):
        content = list_content.getvalue()
        print(content)


class ListStrategyInterface(ABC):
    @abstractmethod
    def start(self, string):
        pass

    @abstractmethod
    def end(self, string):
        pass

    @abstractmethod
    def addListItem(self, string, item):
        pass


class MarkdownListStrategy(ListStrategyInterface):
    def start(self, string):
        pass

    def end(self, string):
        pass

    def addListItem(self, string, item):
        string.write(' * {}\n'.format(item))


class HtmlListStrategy(ListStrategyInterface):
    def start(self, string):
        string.write('<ul>\n')

    def end(self, string):
        string.write('</ul>\n')

    def addListItem(self, string, item):
        string.write(' <li> {} </li>\n'.format(item))

def main():
    text_processor = TextProcessor(MarkdownListStrategy())
    text_processor.append_list(['foo', 'Yao', 'Ling'])
    text_processor.output()

    text_processor.reset()
    text_processor.set_list_strategy(HtmlListStrategy())
    text_processor.append_list(['100', '5K', '$999'])
    text_processor.output()


if __name__ == '__main__':
    main()
